### PR TITLE
Few minor changes to the repo to make it look better

### PR DESCRIPTION
**Pull request by GawdOfROFL**
nothing more than just making it have a more of a modular approach, added argparse for argument parsing in the cli, didnt add any new rule, just moved all the rules in the base directory to a new directory to make the repo look clean thats all. awsome project!

### CODE REVIEW

**Comment by GawdOfROFL** [comment]
@googlebot I signed it!

File: generator.py [hunk 1]
```diff
@@ -368,106 +110,40 @@ def generate_new_sample(template, htmlgrammar, cssgrammar, jsgrammar):
...
-
-    f = open(os.path.join(grammar_dir, 'template.html'))
-    template = f.read()
-    f.close()
-
+def import_grammar():
+    
     htmlgrammar = Grammar()
-    err = htmlgrammar.parse_from_file(os.path.join(grammar_dir, 'html.txt'))
+    err = htmlgrammar.parse_from_file(os.path.join(os.path.dirname(__file__), 'rules/html.txt'))
```
Comment (ifratric) [inline generator.py:116]: Why did you remove the grammar_dir argument? Instead of doing this you could have just set that argument to `os.path.join(os.path.dirname(__file__), 'rules'`) in the caller. I think that'd be cleaner.

File: generator.py [hunk 2]
```diff
@@ -368,106 +110,40 @@ def generate_new_sample(template, htmlgrammar, cssgrammar, jsgrammar):
...
-
-if __name__ == '__main__':
-    main()
+def generate_samples(template):
+    """Generates a set of samples and writes them to the output files.
+    Args:
+      grammar_dir: directory to load grammar files from.
+      outfiles: A list of output filenames.
+    """
+    return generate_new_sample(template, *import_grammar())
```
Comment (ifratric) [inline generator.py:149]: Nit: Missing newline at the end of file

File: generator.py [hunk 3]
```diff
@@ -368,106 +110,40 @@ def generate_new_sample(template, htmlgrammar, cssgrammar, jsgrammar):
...
-    else:
-        print('Arguments missing')
-        print("Usage:")
-        print("\tpython generator.py <output file>")
-        print("\tpython generator.py --output_dir <output directory> --no_of_files <number of output files>")
-
-
-if __name__ == '__main__':
-    main()
+def generate_samples(template):
```
Comment (ifratric) [inline generator.py:143]: This is called generate_samples but in reality only generates a single sample

File: main.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+            outfiles = []
+            for i in range(nsamples):
+                outfiles.append(os.path.join(out_dir, 'fuzz-' + str(i).zfill(5) + '.html'))
+
+            for outfile in outfiles:
+                print('Writing a sample to ' + outfile)
+                try:
+                    with open(outfile, 'w') as f:
+                        result = generator.generate_samples(template)
```
Comment (ifratric) [inline main.py:56]: Previously, when generating multiple samples, the grammars were only loaded once and then used to generate all the samples. This is why generating multiple samples at once was much more efficient than generating one sample at a time. With this PR, the grammars will be loaded anew for *every* new sample. This is going to be much less efficient. I think this is the biggest issue with the PR currently.

File: html_tags.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    'tr': 'HTMLTableRowElement',
+    'track': 'HTMLTrackElement',
+    'tt': 'HTMLUnknownElement',
+    'u': 'HTMLUnknownElement',
+    'ul': 'HTMLUListElement',
+    'var': 'HTMLUnknownElement',
+    'video': 'HTMLVideoElement',
+    'wbr': 'HTMLUnknownElement',
+    'xmp': 'HTMLUnknownElement'
+}
```
Comment (ifratric) [inline html_tags.py:137]: Nit: Missing newline at the end of the file.

File: svg_tags.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    'text': 'SVGTextElement',
+    'textPath': 'SVGTextPathElement',
+    'title': 'SVGTitleElement',
+    'tref': 'SVGElement',
+    'tspan': 'SVGTSpanElement',
+    'unknown': 'SVGElement',
+    'use': 'SVGUseElement',
+    'view': 'SVGViewElement',
+    'vkern': 'SVGElement'
+}
```
Comment (ifratric) [inline svg_tags.py:90]: Nit: Missing newline at the end of the file.

**Review by ifratric — COMMENTED** [review]
Hey, thanks! I left some comments, please see inline.

**Comment by GawdOfROFL** [comment]
did few changes according to the comments, please have a look again, Thanks!

File: generator.py [hunk 7]
```diff
@@ -348,16 +109,16 @@ def generate_new_sample(template, htmlgrammar, cssgrammar, jsgrammar):
         lambda match: add_html_ids(match, htmlctx),
         html
     )
-    generate_html_elements(htmlctx, _N_ADDITIONAL_HTMLVARS)
+    generate_html_elements(htmlctx, 5)
```
Comment (ifratric) [inline generator.py:112]: Not sure why you made this change (as well as 2 other similar changes in this function). It's cleaner to have these as constants rather than hardcoded numbers.

File: html_tags.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
+_HTML_TYPES = {
```
Comment (ifratric) [inline html_tags.py:1]: Missing license header

File: main.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+
+import generator
+import argparse
+import os
+
+def get_parser():
```
Comment (ifratric) [inline main.py:24]: please rename to `get_argument_parser` as "parser" in Domato can mean many things.

File: main.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
+#   Domato - main script
```
Comment (ifratric) [inline main.py:1]: I thought about this some more and I actually prefer to have generator.py serve as main. Please merge it back.

File: svg_tags.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
+_SVG_TYPES = {
```
Comment (ifratric) [inline svg_tags.py:1]: Missing license header.

File: generator.py [hunk 12]
```diff
@@ -368,106 +129,51 @@ def generate_new_sample(template, htmlgrammar, cssgrammar, jsgrammar):
 
     return result
 
-
-def generate_samples(grammar_dir, outfiles):
+def generate_sample(template, outfiles=None):
```
Comment (ifratric) [inline generator.py:132]: This function either returns the generated content or returns nothing and outputs to a file. I don't think that's very clean. If the caller wants to generate a single file they can just pass that single file in the `outfiles` array. No need to implement a separate behavior for generating a single file.

**Review by ifratric — COMMENTED** [review]
Please see follow-up comments inline.

**Comment by GawdOfROFL** [comment]
PTAL

File: generator.py [hunk 13]
```diff
@@ -408,66 +167,74 @@ def generate_samples(grammar_dir, outfiles):
...
 
     for outfile in outfiles:
-        result = generate_new_sample(template, htmlgrammar, cssgrammar,
-                                     jsgrammar)
-
-        if result is not None:
-            print('Writing a sample to ' + outfile)
-            try:
-                f = open(outfile, 'w')
+        result = generate_new_sample(template, htmlgrammar, cssgrammar, jsgrammar)
```
Comment (ifratric) [inline generator.py:170]: Why did you remove the `if result is not None:` check?

File: generator.py [hunk 14]
```diff
@@ -408,66 +167,74 @@ def generate_samples(grammar_dir, outfiles):
...
 
-        outfiles = []
-        for i in range(nsamples):
-            outfiles.append(os.path.join(out_dir, 'fuzz-' + str(i).zfill(5) + '.html'))
+    parser = get_argument_parser()
+    
+    args = parser.parse_args()
 
-        generate_samples(fuzzer_dir, outfiles)
+    if args.file:
```
Comment (ifratric) [inline generator.py:206]: Currently this makes the behavior different than the documentation (according to the readme 'python generator.py <output file>' should work, but it doesn't anymore). Either the documentation needs to change to match the behavior or the other way around.

**Comment by ifratric** [comment]
I added some follow-up comments. Please take a look.

File: generator.py (hunk 13 above)
Comment (GawdOfROFL) [reply generator.py:170]: The reason being that there is almost no possibility of returning None with the current template or if any custom template is being used , as long as it follows the documentation, the generator shouldnt generate a None result, im adding that back in the next commit if you want

File: generator.py (hunk 14 above)
Comment (GawdOfROFL) [reply generator.py:206]: mind if instead of giving the whole usage information i do something like 
```
#### Usage

To see the usage information run the following command:
`python3 generator.py --help`
```
or do i have to just add the `--file option` in the docs?

File: generator.py [hunk 15]
```diff
@@ -368,38 +131,34 @@ def generate_new_sample(template, htmlgrammar, cssgrammar, jsgrammar):
...
     Args:
       grammar_dir: directory to load grammar files from.
       outfiles: A list of output filenames.
     """
 
-    f = open(os.path.join(grammar_dir, 'template.html'))
-    template = f.read()
-    f.close()
-
+    grammar_dir = os.path.dirname(__file__)
```
[the remark this answers is not in the document]
Comment (GawdOfROFL) [reply generator.py:141]: yep thats a good idea :+1: , adding that as well in the next commit

File: generator.py [hunk 16]
```diff
@@ -408,66 +167,74 @@ def generate_samples(grammar_dir, outfiles):
...
-        print('Arguments missing')
-        print("Usage:")
-        print("\tpython generator.py <output file>")
-        print("\tpython generator.py --output_dir <output directory> --no_of_files <number of output files>")
+        parser.print_help()
 
 
 if __name__ == '__main__':
+    
+    _N_MAIN_LINES = 1000
```
[the remark this answers is not in the document]
Comment (GawdOfROFL) [reply generator.py:235]: will there be any possibility of using the generator as module instead of using it as the main script?, and if so why would someone do that? like any particular use case?

[the remark this answers is not in the document]
Comment (ifratric) [reply generator.py:235]: I'd prefer if people could do that if they wanted to, rather than stuff breaking if they do. For example, someone might want to use generate_new_sample in some other way and have their own "main" script.

File: generator.py (hunk 13 above)
Comment (ifratric) [reply generator.py:170]: It appears you are right and it can't return None (maybe it could before at some point).

File: generator.py (hunk 14 above)
Comment (ifratric) [reply generator.py:206]: I like to have a "quick start" command in the documentation. But feel free to include the above as well.

**Comment by GawdOfROFL** [comment]
Pushed the changes , please take a look

**Comment by ifratric** [comment]
LGTM, thanks!
